### PR TITLE
A0-1499: E2e ban test threshold

### PR DESCRIPTION
**Pull request by maciejzelaszczyk**
# Description

This introduces the threshold test for the ban functionality, 4. here:

https://cardinal-cryptography.atlassian.net/browse/A0-1432

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have added tests
- I have created new documentation
- I have bumped aleph-client version if relevant

### CODE REVIEW

File: e2e-tests/src/ban.rs [hunk 1]
```diff
@@ -124,3 +138,58 @@ pub fn check_ban_event<C: AnyConnection>(
...
 }
+
+pub fn check_session_count<C: AnyConnection>(
+    connection: &C,
+    seats: &CommitteeSeats,
+    reserved_validators: &[AccountId],
+    non_reserved_validators: &[AccountId],
+    start_session: &SessionIndex,
+    sessions_to_check: &SessionCount,
+    ban_session_threshold: &SessionCount,
```
Comment (kostekIV) [inline e2e-tests/src/ban.rs:149]: no need for borrows here

File: e2e-tests/src/ban.rs [hunk 2]
```diff
@@ -124,3 +138,58 @@ pub fn check_ban_event<C: AnyConnection>(
...
+) -> anyhow::Result<()> {
+    let validators: Vec<_> = reserved_validators
+        .iter()
+        .chain(non_reserved_validators.iter())
+        .collect();
+
+    let mut expected_validator_session_count = HashMap::new();
+
+    for session in *start_session..*start_session + *sessions_to_check {
+        wait_for_at_least_session(connection, session)?;
```
Comment (kostekIV) [inline e2e-tests/src/ban.rs:159]: rather than waiting here, run this function after `start_session + sessions_to_check` have passed

File: e2e-tests/src/ban.rs [hunk 3]
```diff
@@ -124,3 +138,58 @@ pub fn check_ban_event<C: AnyConnection>(
...
+            if members_for_session.contains(&val) {
+                *session_count += 1;
+                *session_count %= ban_session_threshold;
+            }
+
+            let expected_session_count = expected_validator_session_count
+                .get(&val)
+                .unwrap_or_else(|| panic!("Missing expected session count for validator {}", val));
+
+            check_underperformed_validator_session_count(connection, val, expected_session_count);
```
Comment (kostekIV) [inline e2e-tests/src/ban.rs:190]: from task/test description it should be enough to just check if the validator that was in the committee was marked as underperformer - that would get rid of maps, modulos and counter greatly simplifying the test

File: e2e-tests/src/ban.rs (hunk 1 above)
Comment (maciejzelaszczyk) [reply e2e-tests/src/ban.rs]: Changed.

File: e2e-tests/src/ban.rs (hunk 2 above)
Comment (maciejzelaszczyk) [reply e2e-tests/src/ban.rs]: I have changed the test logic to check the storage values once all the checked sessions have passed.

File: e2e-tests/src/ban.rs (hunk 3 above)
Comment (maciejzelaszczyk) [reply e2e-tests/src/ban.rs]: This is also changed. There is still some less-than-pretty conditional wrangling since the diff in counts between sessions depends on whether the particular validator is in committee and, additionally, the counter might be reset.

File: pallets/elections/src/lib.rs [hunk 4]
```diff
@@ -539,6 +538,8 @@ pub mod pallet {
                 }
             }
 
+            Self::emit_fresh_bans_event();
```
Comment (kostekIV) [inline pallets/elections/src/lib.rs:541]: if this works it should be put in separate pr (with versions, spec_version etc bumped)

**Review by kostekIV — APPROVED** [review]

File: pallets/elections/src/lib.rs (hunk 4 above)
Comment (maciejzelaszczyk) [reply pallets/elections/src/lib.rs]: It is out of this PR, then.

**Review by Marcin-Radecki — APPROVED** [review]
Superb!
